internal/component/loki/secretfilter/secretfilter.go

```diff
@@ -0,0 +1,321 @@
+			}
+			allowlist = append(allowlist, AllowRule{Regex: re, Source: fmt.Sprintf("rule %s", rule.ID)})
+		}
+
+		newRule := Rule{
+			name:        rule.ID,
+			regex:       re,
+			secretGroup: rule.SecretGroup,
+			description: rule.Description,
+			allowlist:   allowlist,
+		}
+
+		// We treat the generic API key rule separately as we want to add it in last position
+		// to the list of rules (so that is has the lowest priority)
+		if strings.ToLower(rule.ID) == "generic-api-key" {
+			ruleGenericApiKey = &newRule
+		} else {
+			c.Rules = append(c.Rules, newRule)
+		}
+	}
+
+	// Compiling global allowlist regexes
+	// From the Gitleaks config
+	for _, r := range gitleaksCfg.AllowList.Regexes {
+		re, err := regexp.Compile(r)
+		if err != nil {
+			level.Error(o.Logger).Log("msg", "error compiling allowlist regex", "error", err)
+			return nil, err
+		}
+		c.AllowList = append(c.AllowList, AllowRule{Regex: re, Source: "gitleaks config"})
+	}
+	// From the arguments
+	for _, r := range args.AllowList {
+		re, err := regexp.Compile(r)
+		if err != nil {
+			level.Error(o.Logger).Log("msg", "error compiling allowlist regex", "error", err)
+			return nil, err
+		}
+		c.AllowList = append(c.AllowList, AllowRule{Regex: re, Source: "alloy config"})
+	}
+
+	// Add the generic API key rule last if needed
+	if ruleGenericApiKey != nil && !args.ExcludeGeneric {
+		c.Rules = append(c.Rules, *ruleGenericApiKey)
+	}
+
+	level.Info(c.opts.Logger).Log("Compiled regexes for secret detection", len(c.Rules))
+
+	// Call to Update() once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	// Immediately export the receiver which remains the same for the component
+	// lifetime.
+	o.OnStateChange(Exports{Receiver: c.receiver})
+
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.receiver.Chan():
+			for _, r := range c.Rules {
+				// To find the secret within the text captured by the regex (and avoid being too greedy), we can use the 'secretGroup' field in the gitleaks.toml file.
+				// But it's rare for regexes to have this field set, so we can use a simple heuristic in other cases.
+				//
+				// There seems to be two kinds of regexes in the gitleaks.toml file
+				// 1. Regexes that only match the secret (with no submatches). E.g. (?:A3T[A-Z0-9]|AKIA|ASIA|ABIA|ACCA)[A-Z0-9]{16}
+				// 2. Regexes that match the secret and some context (or delimiters) and have one submatch (the secret itself). E.g. (?i)\b(AIza[0-9A-Za-z\\-_]{35})(?:['|\"|\n|\r|\s|\x60|;]|$)
+				//
+				// For the first case, we can replace the entire match with the redaction string.
+				// For the second case, we can replace the first submatch with the redaction string (to avoid redacting something else than the secret such as delimiters).
+				for _, occ := range r.regex.FindAllStringSubmatch(entry.Line, -1) {
```

Comment:
Think we can split this block into a separate func, lots of deep nesting going on.

Reply:
Changed in d8fa6e58b10bc03fb912eacbf689cbdf7be45ade.